ILM: make the unfollow action and CCR related steps retryable

andreidan opened the pull request:
This makes the `unfollow` action and the CCR related steps 
retryable.

We already test the retryability infrastructure in several integration tests
in TimeseriesLifecycleIT (ie. testILMRolloverOnManuallyRolledIndex,
testILMRolloverRetriesOnReadOnlyBlock, testRolloverStepRetriesUntilRolledOverIndexIsDeleted,
testSetSingleNodeAllocationRetriesUntilItSucceeds etc.) so this PR doesn't
contain additional tests.

Relates to #48183

andreidan committed:
ILM: make the unfollow action and CCR related steps retryable

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/PauseFollowerIndexStep.java:
```diff
@@ -17,6 +17,11 @@
+
     @Override
     void innerPerformAction(String followerIndex, Listener listener) {
         PauseFollowAction.Request request = new PauseFollowAction.Request(followerIndex);
```
Is there any way that we can check whether the pause has already happened and skip it here? Here's the scenario I'm concerned about:

- Pause gets executed
- Request is not acknowledged
- Assertion is thrown (I think I've seen users run with assertions enabled before)
- Index enters the ERROR state
- Retries the pause
- Because the pause has already occurred, the `TransportPauseFollowAction` throws an exception as pausing twice throws an exception
- Now we're stuck in an ERROR loop

I think we should also handle the case where the index is manually paused and then the `pause-follower-index` step runs, it should skip the index.

andreidan replied:
Yes, ++ on making them idempotent. I'll do that as part of this PR. 

I do wonder with regards to the `isAcked` check. I don't think it should be an assertion as it's something that can happen, so IMO, it should either be an exception or we shouldn't care about it? Given a majority of users run without enabled assertions I'd say we shouldn't care about it (also, an unacked operation is not a failed one). I believe that if we want to wait for all nodes (well, shards in this case) to ack an operation, we should have a step that waits for this condition. 

What do you think?

dakrone replied:
I think it's mostly an assertion for our tests, so that the test will fail and we'll be able to figure out why. Especially in the case of an integration test, we would always want to know if something wasn't acknowledged.

I think I'm good with keeping it, I know that most users don't run with assertions enabled (and I think we don't recommend that they do). The example above is pretty contrived (the assertion one), and I don't think it is important enough for us to need to remove the assertion.

andreidan replied:
@dakrone 
Re:acknowledged /unacknowledged - I agree we'd want to surface this situation if it's a problem. I guess we might view differently an assertion (which I view as expressing/sanity checking "something that should never happen"). A successful unacked operation is something that can happen quite often (?) (especially with larger cluster and many shards), so what I'm proposing is making this explicit (especially now that it'll be a retryable step) and explicitly fail the step if the response is unacked (regardless of assertions enabled/disabled).

Re:idempotency. On a closer look, both `UnfollowFollowerIndexStep` and `PauseFollowerIndexStep` are already idempotent as the `AbstractUnfollowIndexStep` makes sure of this:
```
Map<String, String> customIndexMetadata = indexMetadata.getCustomData(CCR_METADATA_KEY);
        if (customIndexMetadata == null) {
            listener.onResponse(true);
            return;
        }
```

On another note, I renamed `UnfollowFollowIndexStep` to `UnfollowFollowerIndexStep` as the former made my brain spagheti :) 

I'll merge this PR as is if you agree with the rename.

dakrone replied:
> what I'm proposing is making this explicit (especially now that it'll be a retryable step) and explicitly fail the step if the response is unacked (regardless of assertions enabled/disabled).

That sounds good to me!

> On a closer look, both UnfollowFollowerIndexStep and PauseFollowerIndexStep are already idempotent as the AbstractUnfollowIndexStep makes sure of this:
> `if (customIndexMetadata == null)` ...

Looking at the `TransportPauseFollowAction` however, pausing an index doesn't actually remove the `CCR_METADATA_KEY` custom metadata. So the `AbstractUnfollowIndexStep` only makes sure that we skip indices that don't have any CCR configuration. If an index *is* a follower, then double unfollowing still throws this error:

https://github.com/elastic/elasticsearch/blob/ce30ac1a8e2a8de45d88993dbb08312aca31af92/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPauseFollowAction.java#L77

(Which is actually after a check of the same kind):

https://github.com/elastic/elasticsearch/blob/ce30ac1a8e2a8de45d88993dbb08312aca31af92/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPauseFollowAction.java#L57-L60

So we still need to do additional checks to see if there are any follow tasks, basically doing this:

https://github.com/elastic/elasticsearch/blob/ce30ac1a8e2a8de45d88993dbb08312aca31af92/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPauseFollowAction.java#L61-L74

and then checking if the list is empty (in which case we can be idempotent and do `listener.onResponse(true, null)` instead of calling the pause API a second time)

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/UnfollowFollowIndexStep.java:
```diff
@@ -23,6 +23,11 @@
+
     @Override
     void innerPerformAction(String followerIndex, Listener listener) {
         UnfollowAction.Request request = new UnfollowAction.Request(followerIndex);
```
Same comment here about skipping if the index is already unfollowed, a double unfollow throws an error so we'd just be retrying forever if it were executed after the index was already unfollowed

dakrone reviewed (approved):
Sorry this took so long to look at, I wanted to manually validate a couple of things. I think we can improve the resilience by making some of these steps idempotent in their actions.

This PR LGTM, as long as we tackle the concerns in a subsequent one (or if we want to fix them in this PR that's fine too)

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-unfollow-retryable

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/OpenFollowerIndexStep.java:
```diff
@@ -20,6 +20,11 @@
     }
 
+    @Override
+    public boolean isRetryable() {
```
Great spot! I'll make this change in a followup

andreidan committed:
Rename UnfollowFollowIndexStep to UnfollowFollowerIndexStep

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-unfollow-retryable

elasticmachine committed:
Merge branch 'master' into ilm-unfollow-retryable

andreidan committed:
Move ShardFollowTask to `xpack.core`

andreidan committed:
AbstractUnfollowIndexStep skips if the managed index is not a follower index

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-unfollow-retryable

andreidan committed:
Make PauseFollowerIndexStep idempotent

This moves the idempotency checkes from AbstractUnfollowIndexStep to
PauseFollowerIndexStep as the UnfollowFollowerIndexStep was already
idemptotent

andreidan committed:
Remove unused import

andreidan committed:
Convert response isAcknoledged checks from assertion to exceptions

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ilm-unfollow-retryable

andreidan committed:
Convert isAcknowledged assertion to exception

dakrone reviewed (approved):
LGTM, thanks for working through this Andrei!
